### ISSUE TITLE
How to make visualization

### ISSUE THREAD

**Issue by Hongxiii** [NVlabs/EAGLE#62]
Hi, thanks for releasing this work.

I can run the model and obtain the detection result, and I can also parse the predicted boxes using `parse_boxes()`. However, I am not sure how to obtain the final visualization result, i.e., drawing the detected bounding boxes and corresponding labels on the original image.

Could you please provide an example or recommended way to visualize the detection/grounding results? For example, how to save an output image with boxes and labels such as `person`, `pet`, or `smiling face`?

Thanks!

**Comment by exiawsh** [comment]
# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.

"""
locateanything_worker.py - A reusable worker for LocateAnything inference.
"""
import re
import os

import torch
from PIL import Image, ImageDraw, ImageFont
from transformers import AutoModel, AutoTokenizer, AutoProcessor


class LocateAnythingWorker:
    """Stateful worker that loads the model once and serves perception queries."""

    def __init__(self, model_path: str, device: str = "cuda", dtype=torch.bfloat16):
        self.device = device
        self.dtype = dtype

        self.tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
        self.processor = AutoProcessor.from_pretrained(model_path, trust_remote_code=True)
        self.model = AutoModel.from_pretrained(
            model_path,
            torch_dtype=dtype,
            trust_remote_code=True,
        ).to(device).eval()

    @torch.no_grad()
    def predict(
        self,
        image: Image.Image,
        question: str,
        generation_mode: str = "hybrid",
        max_new_tokens: int = 2048,
        temperature: float = 0.7,
        verbose: bool = True,
    ) -> dict:
        """
        Run a single perception query.

        Args:
            image: PIL Image (RGB).
            question: The task prompt (see supported prompts below).
            generation_mode: "fast" (MTP) | "slow" (NTP) | "hybrid".
            max_new_tokens: Maximum tokens to generate.
            temperature: Sampling temperature (0 = greedy).
            verbose: If True, return timing statistics.

        Returns:
            dict with keys: "answer", "stats" (optional), "history" (optional).
        """
        messages = [
            {"role": "user", "content": [
                {"type": "image", "image": image},
                {"type": "text", "text": question},
            ]}
        ]

        text = self.processor.py_apply_chat_template(
            messages, tokenize=False, add_generation_prompt=True
        )
        images, videos = self.processor.process_vision_info(messages)
        inputs = self.processor(
            text=[text], images=images, videos=videos, return_tensors="pt"
        ).to(self.device)

        pixel_values = inputs["pixel_values"].to(self.dtype)
        input_ids = inputs["input_ids"]
        image_grid_hws = inputs.get("image_grid_hws", None)

        response = self.model.generate(
            pixel_values=pixel_values,
            input_ids=input_ids,
            attention_mask=inputs["attention_mask"],
            image_grid_hws=image_grid_hws,
            tokenizer=self.tokenizer,
            max_new_tokens=max_new_tokens,
            use_cache=True,
            generation_mode=generation_mode,
            temperature=temperature,
            do_sample=True,
            top_p=0.9,
            repetition_penalty=1.1,
            verbose=verbose,
        )

        result = {"answer": response[0] if isinstance(response, tuple) else response}
        if isinstance(response, tuple) and len(response) >= 3:
            result["history"] = response[1]
            result["stats"] = response[2]
        return result

    # ---- Convenience methods for each task ----

    def detect(self, image: Image.Image, categories: list[str], **kwargs) -> dict:
        """Object detection / document layout analysis."""
        cats = "</c>".join(categories)
        prompt = f"Locate all the instances that matches the following description: {cats}."
        return self.predict(image, prompt, **kwargs)

    def ground_single(self, image: Image.Image, phrase: str, **kwargs) -> dict:
        """Phrase grounding — single instance."""
        prompt = f"Locate a single instance that matches the following description: {phrase}."
        return self.predict(image, prompt, **kwargs)

    def ground_multi(self, image: Image.Image, phrase: str, **kwargs) -> dict:
        """Phrase grounding — multiple instances."""
        prompt = f"Locate all the instances that match the following description: {phrase}."
        return self.predict(image, prompt, **kwargs)

    def ground_text(self, image: Image.Image, phrase: str, **kwargs) -> dict:
        """Text grounding."""
        prompt = f"Please locate the text referred as {phrase}."
        return self.predict(image, prompt, **kwargs)

    def detect_text(self, image: Image.Image, **kwargs) -> dict:
        """Scene text detection."""
        prompt = "Detect all the text in box format."
        return self.predict(image, prompt, **kwargs)

    def ground_gui(self, image: Image.Image, phrase: str, output_type: str = "box", **kwargs) -> dict:
        """GUI grounding (box or point)."""
        if output_type == "point":
            prompt = f"Point to: {phrase}."
        else:
            prompt = f"Locate the region that matches the following description: {phrase}."
        return self.predict(image, prompt, **kwargs)

    def point(self, image: Image.Image, phrase: str, **kwargs) -> dict:
        """Pointing."""
        prompt = f"Point to: {phrase}."
        return self.predict(image, prompt, **kwargs)

    # ---- Utility: parse model output ----

    @staticmethod
    def parse_boxes(answer: str, image_width: int, image_height: int) -> list[dict]:
        """Parse model output into pixel-coordinate bounding boxes, including labels if present.
        
        Handles output like: <ref>person</ref><box><100><200><300><400></box>
        Coordinates in model output are normalized integers in [0, 1000].
        """
        boxes = []
        # Pattern to find optional <ref> tags followed by <box> tags
        pattern = r"(?:<ref>(.*?)</ref>\s*)?<box><(\d+)><(\d+)><(\d+)><(\d+)></box>"
        
        for m in re.finditer(pattern, answer):
            label = m.group(1) if m.group(1) else "object"
            x1, y1, x2, y2 = [int(g) for g in m.groups()[1:]]
            
            boxes.append({
                "label": label,
                "x1": x1 / 1000 * image_width,
                "y1": y1 / 1000 * image_height,
                "x2": x2 / 1000 * image_width,
                "y2": y2 / 1000 * image_height,
            })
        return boxes

    @staticmethod
    def parse_points(answer: str, image_width: int, image_height: int) -> list[dict]:
        """Parse model output into pixel-coordinate points, including labels."""
        points = []
        pattern = r"(?:<ref>(.*?)</ref>\s*)?<box><(\d+)><(\d+)></box>"
        
        for m in re.finditer(pattern, answer):
            label = m.group(1) if m.group(1) else "point"
            x, y = int(m.group(2)), int(m.group(3))
            points.append({
                "label": label,
                "x": x / 1000 * image_width,
                "y": y / 1000 * image_height,
            })
        return points

    # ---- Utility: Visualization ----

    @staticmethod
    def get_color_for_label(label: str):
        """Assigns a consistent color based on the label string."""
        colors = [
            (8, 145, 178), (220, 38, 38), (22, 163, 74), (37, 99, 235),
            (217, 119, 6), (147, 51, 234),
        ]
        idx = sum(ord(c) for c in label)
        return colors[idx % len(colors)]

    @staticmethod
    def draw_visualizations(image: Image.Image, parsed_results: list[dict], output_path: str = None) -> Image.Image:
        """
        Draws bounding boxes and/or points with labels on an image.
        
        Args:
            image: Original PIL Image.
            parsed_results: List of parsed dictionaries from parse_boxes or parse_points.
            output_path: If provided, saves the visualized image to this path.
            
        Returns:
            The visualized PIL Image.
        """
        pil_img = image.convert("RGBA")
        overlay = Image.new("RGBA", pil_img.size, (255, 255, 255, 0))
        draw = ImageDraw.Draw(overlay)
        
        try:
            # Fallback to default font if arial is missing
            font = ImageFont.truetype("arial.ttf", 20)
        except IOError:
            font = ImageFont.load_default()

        w_img, h_img = pil_img.size

        for res in parsed_results:
            label = res.get("label", "object")
            color = LocateAnythingWorker.get_color_for_label(label)
            
            # Handle Box Drawing
            if "x1" in res and "x2" in res:
                x1, y1 = max(0, res["x1"]), max(0, res["y1"])
                x2, y2 = min(w_img, res["x2"]), min(h_img, res["y2"])
                
                fill_color = color + (65,) # RGBA with 65 alpha
                draw.rectangle([x1, y1, x2, y2], fill=fill_color, outline=color, width=4)
                
                if label:
                    t_box = draw.textbbox((0, 0), label, font=font)
                    tw = t_box[2] - t_box[0]
                    th = t_box[3] - t_box[1]
                    
                    pad_x, pad_y = 7, 4
                    tag_w, tag_h = tw + pad_x * 2, th + pad_y * 2
                    tag_y = y1 - tag_h - 2
                    
                    if tag_y < 0:
                        tag_y = y2 + 2
                        
                    draw.rectangle([x1, tag_y, x1 + tag_w, tag_y + tag_h], fill=color)
                    draw.text((x1 + pad_x, tag_y + pad_y), label, fill="white", font=font)
                    
            # Handle Point Drawing
            elif "x" in res and "y" in res:
                cx, cy = res["x"], res["y"]
                r = 10
                draw.ellipse([cx - r, cy - r, cx + r, cy + r], fill=color, outline="white", width=2)
                
                if label:
                    t_box = draw.textbbox((0, 0), label, font=font)
                    th, tw = t_box[3] - t_box[1], t_box[2] - t_box[0]
                    tx, ty = cx + 14, cy - th // 2
                    
                    draw.rectangle([tx - 2, ty - 2, tx + tw + 6, ty + th + 4], fill=color)
                    draw.text((tx + 2, ty), label, fill="white", font=font)

        final_img = Image.alpha_composite(pil_img, overlay).convert("RGB")
        
        if output_path:
            final_img.save(output_path)
            
        return final_img


# --------------- Usage Example ---------------
if __name__ == "__main__":
    worker = LocateAnythingWorker("nvidia/LocateAnything-3B")
    
    # Ensure there is an example.jpg in the directory for testing
    img = Image.open("example.jpg").convert("RGB")
    w, h = img.size

    # 1. Object Detection
    print("--- Object Detection ---")
    result = worker.detect(img, ["person", "car", "bicycle"])
    print("Raw Answer:", result["answer"])
    
    # Parse the results and generate the visualization
    boxes = worker.parse_boxes(result["answer"], w, h)
    worker.draw_visualizations(img, boxes, "output_detection.jpg")
    print("Saved visualization to output_detection.jpg")

    # 2. Pointing
    print("\n--- Pointing ---")
    result = worker.point(img, "the traffic light")
    print("Raw Answer:", result["answer"])
    
    # Parse the point coordinates and generate the visualization
    points = worker.parse_points(result["answer"], w, h)
    worker.draw_visualizations(img, points, "output_pointing.jpg")
    print("Saved visualization to output_pointing.jpg")

@Hongxiii Thanks for your attention! Try this
